### DISCUSSION TITLE
Problem with datachange notification

### DISCUSSION THREAD

**Discussion by Statkern15** [FreeOpcUa/opcua-asyncio#1685] [Q&A]
Hello, 
I have a strange behaviour of my subscriptions. When subscribing to multiple nodes (around 10) the subscription works as expected -> generating datachange_notification when the value has changed. But after some time (and this can vary from a couple of hours to days) I get one data_notification for the same MonitoredItem where the value is the same but the timestamp has changed slightly. 
Here is a shorten example of such a notification 
```    
MonitoredItemNotification(ClientHandle=210, Value=DataValue(Value=Variant(Value=2.7250001430511475, VariantType=<VariantType.Float: 10>, Dimensions=None, is_array=False), StatusCode_=StatusCode(value=0), 
SourceTimestamp=datetime.datetime(2024, 7, 24, 15, 26, 16, 130329, tzinfo=datetime.timezone.utc), ServerTimestamp=datetime.datetime(2024, 7, 24, 15, 26, 16, 130329, tzinfo=datetime.timezone.utc), SourcePicoseconds=None, ServerPicoseconds=None)), 

MonitoredItemNotification(ClientHandle=222, Value=DataValue(Value=Variant(Value=2.7250001430511475, VariantType=<VariantType.Float: 10>, Dimensions=None, is_array=False), StatusCode_=StatusCode(value=0), 
SourceTimestamp=datetime.datetime(2024, 7, 24, 15, 26, 16, 130743, tzinfo=datetime.timezone.utc), ServerTimestamp=datetime.datetime(2024, 7, 24, 15, 26, 16, 130743, tzinfo=datetime.timezone.utc), SourcePicoseconds=None, ServerPicoseconds=None)),

MonitoredItemNotification(ClientHandle=234, Value=DataValue(Value=Variant(Value=2.7250001430511475, VariantType=<VariantType.Float: 10>, Dimensions=None, is_array=False), StatusCode_=StatusCode(value=0), 
SourceTimestamp=datetime.datetime(2024, 7, 24, 15, 26, 16, 131837, tzinfo=datetime.timezone.utc), ServerTimestamp=datetime.datetime(2024, 7, 24, 15, 26, 16, 131837, tzinfo=datetime.timezone.utc), SourcePicoseconds=None, ServerPicoseconds=None)), 
```    
The subscription is configure with the publishing Interval set to 500ms. And this is my code so far: 
``` 
try:
    async with self.client:
        handler = sub_handler
        subscription = await self.client.create_subscription(500, handler)
        await subscription.subscribe_data_change(nodes)
        while True:
            await asyncio.sleep(5)
            await self.client.check_connection()  # Throws a exception if connection is lost
except ....
``` 
I have no idea why this happens and if its a client issue or a server issue although this happens on multiple different servers. If i restart the client this behaviour disapears (at least for a while) which indicates that it must come from the client. 
Has someone had a similar problem or is there a way to better debug this behaviour.

Thanks in advance

**Comment by oroulet** [comment]
It is always hard to be sure but due to the way things are implemented I think this is a sever issue. On the client side we just forward to python what we receive from server. We never create an event on our own

**Reply by Statkern15** [reply]
Thanks for the reply. This is bad news since I dont't have controll over the server. 
Is there a easy way to filter those "dublicates" if they happen within a defines timespan?

**Comment by Statkern15** [comment]
Ok, I think I might have found the issue, and it looks to be a client issue.
In order to enable reconnecting, I based my code on the example here: https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/client-reconnect.py.

As it looks, when an exception is thrown, my client does not handle the cleanup correctly, meaning when reconnecting, it adds additional monitored items to the subscription. So, if it connects for the first time and adds, for example, 10 items, when reconnecting, it adds another 10, leaving me with 20 monitored items in total. This explains why this only happens over time and disappears when restarting the client.
My initial thought was that the monitored items are lost when the connection is lost, but that is apparently not the case.
I will add a cleanup and hope that this will solve the issue.
